### PR TITLE
Bim/feature/33281 table display mode

### PR DESCRIPTION
**Pull request by aleixsuau**
https://community.openproject.com/projects/bim/work_packages/33281

The main goals of this pull request are:

- [x] Add a new "Table display" mode to the BimViewToggleDropdownDirective

- [x]  Add the bcfThumbnail column to the table on the  BcfListContainerComponent

- [x] Add tests for the previous behaviors.

The temporary approach followed to reach the first goal has been passing an extra parameter 'cards' to the router. The BcfListContainerComponent listens to the this.uIRouterGlobals.params$ to switch between cards and list view. Final solution is pending of approval.

### CODE REVIEW

**Comment by wielinde** [comment]
Hey @aleixsuau, Travis doesn't pass as multiple specs are failing and the look like they fails due to changes in the PR.

File: config/middleware.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+require 'request_headers'
```
Comment (wielinde) [inline config/middleware.rb]: Remove this file.

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  background-color: #FFFFFF
+  opacity: 0.5
+
+.work-packages--card-view-container
+  width: 100%
+  overflow: auto
+  padding-bottom: 5px
+  @include styled-scroll-bar
+
+  @media screen and (min-width: 679px)
```
Comment (wielinde) [inline frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass:14]: These min-width values looks to me as if they should be variables instead. What is this breakpoint? Can we give it a name?

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    &.-with-resizer
+     padding-left: 10px
+
+  @media screen and (max-width: 679px)
```
Comment (wielinde) [inline frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass:18]: Again, can we use a variable here?

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    // Ensure the WP cards span the complete width on mobile
+    // --> Move scrollbar out of visible area
+    min-width: calc(100vw + 10px)
+
+.work-packages-split-view--tabletimeline-container
+  display: flex
+  flex: 1 1 auto
+  @include styled-scroll-bar
+
+  @media screen and (min-width: 679px)
```
Comment (wielinde) [inline frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass:28]: again

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    &.-with-resizer
+     padding-left: 10px
+
+  @media screen and (max-width: 679px)
```
Comment (wielinde) [inline frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass:32]: again

File: frontend/src/app/modules/work_packages/routing/wp-list-view/wp-list-view.component.sass [hunk 6]
```diff
@@ -19,3 +19,17 @@
     // Ensure the WP cards span the complete width on mobile
     // --> Move scrollbar out of visible area
     min-width: calc(100vw + 10px)
+
+.work-packages-split-view--tabletimeline-container
+  display: flex
+  flex: 1 1 100%
+  @include styled-scroll-bar
+
+  @media screen and (min-width: 679px)
```
Comment (wielinde) [inline frontend/src/app/modules/work_packages/routing/wp-list-view/wp-list-view.component.sass:28]: again

File: frontend/src/app/modules/work_packages/routing/wp-list-view/wp-list-view.component.sass [hunk 7]
```diff
@@ -19,3 +19,17 @@
...
+    &.-with-resizer
+      padding-left: 10px
+
+  @media screen and (max-width: 679px)
```
Comment (wielinde) [inline frontend/src/app/modules/work_packages/routing/wp-list-view/wp-list-view.component.sass:32]: again

File: spec/support/capybara_headers.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+require 'request_headers'
```
Comment (wielinde) [inline spec/support/capybara_headers.rb]: remove this file

**Review by wielinde — CHANGES_REQUESTED** [review]
Hey @aleixsuau, the PR looks pretty good. However, there is one important feature missing or wrong. - * 
- When in "viewer and table" mode and
- I click on a WP ID
- Then unexpectedly the whole page reloads
- Instead I would expect that it only replaces the WP list with the WP details just as a double click on a card.

File: spec/support/capybara_headers.rb (hunk 8 above)
Comment (wielinde) [reply spec/support/capybara_headers.rb]: Also the header of the table doesn't look aligned with the toolbar of the viewer. I know that they technically are, but is looks ugly.
image

**Comment by wielinde** [comment]
Further, the info icon at the end of a table row links away from the viewer. I would expect that it either splits the space of the list into another split, just as the Gantt chart does, or that it behaves the same as the click on the info icon on a card.

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass (hunk 2 above)
Comment (aleixsuau) [reply frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass:14]: @wielinde Previously this component shared template with the wp-list-view.component.ts, I just copied its template and styles and applied some modifications. These styles were already there so I'm not sure.

I see this same breakpoint used multiple times on the project so it seems a good idea to replace it with a variable, but maybe in its own pull request?

**Comment by aleixsuau** [comment]
> Hey @aleixsuau, the PR looks pretty good. However, there is one important feature missing or wrong. - *
> 
> * When in "viewer and table" mode and
> * I click on a WP ID
> * Then unexpectedly the whole page reloads
> * Instead I would expect that it only replaces the WP list with the WP details just as a double click on a card.

> Further, the info icon at the end of a table row links away from the viewer. I would expect that it either splits the space of the list into another split, just as the Gantt chart does, or that it behaves the same as the click on the info icon on a card.

Hi @wielinde,

Thanks for your comments. I'm aware of these two behaviors. The underlying cause of them is that the table handles both events (clicking on the wp id and on the info icon) with its own default actions (TableHandlerRegistry). After talking with @oliverguenther we decided to create an @Output / @Input interface for the table that will allow reacting to them in a custom way from a parent component (ie: <table (wpClicked)="onWpClicked($event)"></table>). This is the work package for it https://community.openproject.com/projects/bim/work_packages/33868.

It is also in the plan to create the same interface for the cards.

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass (hunk 2 above)
Comment (wielinde) [reply frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.sass:14]: OK with me.

**Comment by aleixsuau** [comment]
Hi @wielinde 👋 
Could you please take a look at this PR?
"Keeping the scroll position when coming back from a detail page" is done on this pull request: https://github.com/opf/openproject/pull/8496

**Comment by aleixsuau** [comment]
> Further, the info icon at the end of a table row links away from the viewer. I would expect that it either splits the space of the list into another split, just as the Gantt chart does, or that it behaves the same as the click on the info icon on a card.

This is fixed with the last changes @wielinde

**Comment by aleixsuau** [comment]
> Hey @aleixsuau, the PR looks pretty good. However, there is one important feature missing or wrong. - *
> 
> * When in "viewer and table" mode and
> * I click on a WP ID
> * Then unexpectedly the whole page reloads
> * Instead I would expect that it only replaces the WP list with the WP details just as a double click on a card.

This is fixed with the last changes @wielinde

**Comment by wielinde** [comment]
@aleixsuau I can't get it compliled:

```
ERROR in src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.ts:26:14 - error TS2415: Class 'BcfListContainerComponent' incorrectly extends base class 'WorkPackageListViewComponent'.
  Property 'showResizerInCardView' is protected in type 'BcfListContainerComponent' but public in type 'WorkPackageListViewComponent'.

26 export class BcfListContainerComponent extends WorkPackageListViewComponent implements OnInit {
                ~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by aleixsuau** [comment]
> @aleixsuau I can't get it compliled:
> 
> ```
> ERROR in src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.ts:26:14 - error TS2415: Class 'BcfListContainerComponent' incorrectly extends base class 'WorkPackageListViewComponent'.
>   Property 'showResizerInCardView' is protected in type 'BcfListContainerComponent' but public in type 'WorkPackageListViewComponent'.
> 
> 26 export class BcfListContainerComponent extends WorkPackageListViewComponent implements OnInit {
>                 ~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

Sorry @wielinde , now it should work.

File: lib/request_headers.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+class CustomHeadersHelper
```
Comment (wielinde) [inline lib/request_headers.rb]: This file should not be here. Grrrr.

File: modules/bim/config/locales/crowdin/js-es.yml [hunk 10]
```diff
@@ -18,4 +18,3 @@ es:
         manage: 'Administrar modelos'
       views:
         viewer: 'Visor'
-        split: 'Visor y tarjetas'
```
Comment (wielinde) [inline modules/bim/config/locales/crowdin/js-es.yml:21]: You should not edit yml translation files that are not the English ones.

**Comment by wielinde** [comment]
> > Hey @aleixsuau, the PR looks pretty good. However, there is one important feature missing or wrong. - *
> > 
> > * When in "viewer and table" mode and
> > * I click on a WP ID
> > * Then unexpectedly the whole page reloads
> > * Instead I would expect that it only replaces the WP list with the WP details just as a double click on a card.
> 
> This is fixed with the last changes @wielinde

Hey @aleixsuau ,

I've just played around with the PR and I am pretty sure, that the default behavior for opening a work package from a table should be different. Currently it opens the WP to the right of the table. I argue that in most cases there is not enough space available on the screen. This is why we also decided in the cards view to replace the cards with the WP and not show the WP on the right side of the cards list. So I am convinced that we should do the same when in table mode.

Just imagine you are in the Revit addin and you don't have a viewer but very little space in your addin window....

**Comment by wielinde** [comment]
@aleixsuau please create separate WP+PR for the comment above about replacing the lists with the full view when in Revit Addin. So if you fix the other two issues above feel free to merge this PR into `dev`.

File: modules/bim/config/locales/crowdin/js-es.yml (hunk 10 above)
Comment (aleixsuau) [reply modules/bim/config/locales/crowdin/js-es.yml:21]: This change is deleting my previous edition (adding "split: 'Visor y tarjetas' ").

File: lib/request_headers.rb (hunk 9 above)
Comment (aleixsuau) [reply lib/request_headers.rb]: Ok, removed

**Comment by wielinde** [comment]
@aleixsuau If travis passes, feel free to merge.

**Pull request merged**